sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -122,75 +201,25 @@ case class GpuGenerateExec(
-
-    val outputSchema = output.map(_.dataType).toArray
-    child.executeColumnar().mapPartitions { it =>
-      new Iterator[ColumnarBatch] {
-        var currentBatch: ColumnarBatch = _
-        var indexIntoData = 0
-
-        private def closeCurrent(): Unit = if (currentBatch != null) {
-          currentBatch.close()
-          currentBatch = null
-        }
-
-        TaskContext.get().addTaskCompletionListener[Unit](_ => closeCurrent())
-
-        def fetchNextBatch(): Unit = {
-          indexIntoData = 0
-          closeCurrent()
-          if (it.hasNext) {
-            currentBatch = it.next()
-          }
-        }
-
-        override def hasNext: Boolean = {
-          if (currentBatch == null || indexIntoData >= numArrayColumns) {
-            fetchNextBatch()
-          }
-          currentBatch != null
-        }
-
-        override def next(): ColumnarBatch = {
-          if (currentBatch == null || indexIntoData >= numArrayColumns) {
-            fetchNextBatch()
-          }
-          withResource(new NvtxWithMetrics("GpuGenerateExec", NvtxColor.PURPLE, totalTime)) { _ =>
-            val result = new Array[ColumnVector](numExplodeColumns + numOtherColumns)
-            try {
-              withResource(GpuProjectExec.project(currentBatch, boundOthersProjectList)) { cb =>
-                (0 until cb.numCols()).foreach { i =>
-                  result(i) = cb.column(i).asInstanceOf[GpuColumnVector].getBase.incRefCount()
-                }
-              }
-              if (includePos) {
-                result(numOtherColumns) = withResource(GpuScalar.from(indexIntoData, IntegerType)) {
-                  scalar => ColumnVector.fromScalar(scalar, currentBatch.numRows())
-                }
-              }
-              result(numOtherColumns + numExplodeColumns - 1) =
-                withResource(GpuProjectExec.project(currentBatch,
-                  Seq(boundArrayProjectList(indexIntoData)))) { cb =>
-                  cb.column(0).asInstanceOf[GpuColumnVector].getBase.incRefCount()
-                }
-
-              withResource(new Table(result: _*)) { table =>
-                indexIntoData += 1
-                numOutputBatches += 1
-                numOutputRows += table.getRowCount
-                GpuColumnVector.from(table, outputSchema)
-              }
-            } finally {
-              result.safeClose()
-            }
+    val generatorProjectList: Seq[GpuExpression] =
+      GpuBindReferences.bindGpuReferences(generator.children, child.output)
+    val othersProjectList: Seq[GpuExpression] =
+      GpuBindReferences.bindGpuReferences(requiredChildOutput, child.output)
+
+    child.executeColumnar().map { inputFromChild =>
+      withResource(inputFromChild) { input =>
+        withResource(new NvtxWithMetrics("GpuGenerateExec", NvtxColor.PURPLE, totalTime)) { _ =>
+          // Project input columns, setting other columns ahead of generator's input columns.
```

Comment:
We lost the optimizations we had in place to reduce total memory used when exploding a literal array.  We need to have a way to put that back in.  Similarly this has no memory optimizations.  It is very easy to have the amount of memory used in an output batch grow to a massive size because of the number of elements in the array/map passed in. At a minimum when we don't know if it is a literal/scalar value passed in for the explode, then we need to have the generator do some estimation to split up the input batch into smaller manageable chunks so we have some confidence that we will be close to the desired output batch size.

Reply:
I put back memory optimization via splitting up the input batch into smaller manageable chunks.
If it is a  literal/scalar value passed in for the explode, which means number of row expansion are fixed, we can figure out number of splits from meta data.
If not, we estimate the number of row expansion with calculating average array size of sampling data.